# CCBD SP26 – Variant 2: Partitioning Strategies
## Analysis Notebook

This notebook uses **native Plotly interactive charts** with built-in dropdowns and buttons.
Charts are fully interactive when saved — filters work on GitHub and nbviewer without any extra dependencies.

> Run `Kernel → Restart & Run All`, then save before pushing to GitHub.

**Metrics analysed:** object count, storage size, S3 listing time, throughput, selective query, broad query.

In [1]:
import pandas as pd
import plotly.graph_objects as go
from plotly.subplots import make_subplots

df = pd.read_csv("results.csv")

SIZES        = ["S", "M", "L"]
LAYOUT_ORDER = ["flat", "by_date", "by_region"]
COLORS       = {"flat": "#4C72B0", "by_date": "#DD8452", "by_region": "#55A868"}
SIZE_ROWS    = {"S": 5_000_000, "M": 25_000_000, "L": 100_000_000}

print("Results loaded.")
df

Results loaded.


,size,layout,local_listing_time_s,local_file_count,local_total_size_gb,s3_listing_time_s,s3_object_count,s3_total_size_gb,upload_throughput_mbs,download_throughput_mbs,selective_time_s,selective_rows,broad_time_s,broad_rows
0,S,flat,0.0009,1,0.0737,0.2763,1,0.0737,159.72,220.44,0.4441,64587,0.2618,5000000
1,S,by_date,0.0771,157,0.1060,2.8482,157,0.1060,11.96,24.23,0.1984,96919,1.5484,5000000
2,S,by_region,0.0119,25,0.1029,0.4410,25,0.1029,29.81,82.88,0.1483,96919,0.1583,5000000
3,M,flat,0.0002,1,0.3686,0.0319,1,0.3686,197.60,206.26,0.7105,322200,0.4359,25000000
4,M,by_date,0.1365,784,0.5298,14.1727,784,0.5298,23.95,46.06,0.2521,483372,1.9132,25000000
5,M,by_region,0.0239,125,0.5145,2.2770,125,0.5145,52.16,89.61,0.2024,483372,0.7193,25000000
6,L,flat,0.0004,1,1.4745,0.0364,1,1.4745,322.21,483.15,2.6843,1291441,1.7625,100000000
7,L,by_date,0.1908,3136,2.1193,29.8681,3136,2.1193,11.68,43.32,1.0302,1936418,8.2011,100000000
8,L,by_region,0.0160,500,2.0580,0.2124,500,2.0580,59.72,203.46,1.0607,1936418,2.6445,100000000


## 1. Object Count and Storage Size

Use the **dropdown** to switch between number of files and storage size.
by_date at size L produces **3,136 files** vs 1 for flat, and is ~44% larger due to per-file Parquet metadata.

In [11]:
metrics = [
    ("local_file_count",    "Number of files",    "{:,.0f}"),
    ("local_total_size_gb", "Storage size (GB)",  "{:.3f}"),
]

fig = go.Figure()

for m_idx, (col, label, fmt) in enumerate(metrics):
    for layout in LAYOUT_ORDER:
        subset = df[df["layout"] == layout].set_index("size").reindex(SIZES)
        vals = subset[col]
        fig.add_trace(go.Bar(
            name=layout,
            x=SIZES,
            y=vals,
            marker_color=COLORS[layout],
            text=[fmt.format(v) for v in vals],
            textposition="outside",
            visible=(m_idx == 0),
            legendgroup=layout,
            showlegend=(m_idx == 0),
            hovertemplate=layout + " %{x}: %{y}<extra></extra>",
        ))

n = len(LAYOUT_ORDER)
buttons = []
for m_idx, (col, label, fmt) in enumerate(metrics):
    visible = [False] * (len(metrics) * n)
    for i in range(n):
        visible[m_idx * n + i] = True
    buttons.append(dict(
        label=label,
        method="update",
        args=[{"visible": visible}, {"yaxis.title.text": label}],
    ))

fig.update_layout(
    barmode="group",
    title_text="Object count and storage size by layout",
    xaxis_title="Dataset size",
    yaxis_title="Number of files",
    height=470,
    legend_title="Layout",
    template="plotly_white",
    updatemenus=[dict(
        type="dropdown",
        x=0.0, y=1.1,
        xanchor="left",
        showactive=True,
        buttons=buttons,
    )],
)
fig.show()

## 2. S3 Listing Time

Use the **buttons** to switch between S3 and local listing.
At size L, by_date takes **33s on S3** vs **0.22s locally** — a 153x overhead from paginated API calls.

In [10]:
listing_metrics = [
    ("s3_listing_time_s",   "S3 listing time (s)"),
    ("local_listing_time_s","Local listing time (s)"),
]

fig = go.Figure()

for m_idx, (col, label) in enumerate(listing_metrics):
    for layout in LAYOUT_ORDER:
        subset = df[df["layout"] == layout].set_index("size").reindex(SIZES)
        fig.add_trace(go.Bar(
            name=layout,
            x=SIZES,
            y=subset[col],
            marker_color=COLORS[layout],
            text=subset[col].round(4),
            textposition="outside",
            visible=(m_idx == 0),
            legendgroup=layout,
            showlegend=(m_idx == 0),
            hovertemplate=layout + " %{x}: %{y:.4f}s<extra></extra>",
        ))

n = len(LAYOUT_ORDER)
buttons = []
for m_idx, (col, label) in enumerate(listing_metrics):
    visible = [False] * (len(listing_metrics) * n)
    for i in range(n):
        visible[m_idx * n + i] = True
    buttons.append(dict(
        label=label,
        method="update",
        args=[{"visible": visible}, {"yaxis.title.text": label}],
    ))

fig.update_layout(
    barmode="group",
    title_text="Listing time — lower is better",
    xaxis_title="Dataset size",
    yaxis_title="S3 listing time (s)",
    height=470,
    legend_title="Layout",
    template="plotly_white",
    updatemenus=[dict(
        type="buttons",
        x=0.0, y=1.1,
        xanchor="left",
        showactive=True,
        buttons=buttons,
    )],
)
fig.show()

## 3. Upload and Download Throughput

Use the **buttons** to switch between upload and download.
Flat achieves **246 MB/s upload** vs **8.3 MB/s for by_date** at size L — a 30x difference.

In [7]:
throughput_metrics = [
    ("upload_throughput_mbs",   "Upload (MB/s)"),
    ("download_throughput_mbs", "Download (MB/s)"),
]

fig = go.Figure()

for m_idx, (col, label) in enumerate(throughput_metrics):
    for layout in LAYOUT_ORDER:
        subset = df[df["layout"] == layout].set_index("size").reindex(SIZES)
        fig.add_trace(go.Bar(
            name=layout,
            x=SIZES,
            y=subset[col],
            marker_color=COLORS[layout],
            text=subset[col].round(1),
            textposition="outside",
            visible=(m_idx == 0),
            legendgroup=layout,
            showlegend=(m_idx == 0),
            hovertemplate=layout + " %{x}: %{y:.1f} MB/s<extra></extra>",
        ))

n = len(LAYOUT_ORDER)
buttons = []
for m_idx, (col, label) in enumerate(throughput_metrics):
    visible = [False] * (len(throughput_metrics) * n)
    for i in range(n):
        visible[m_idx * n + i] = True
    buttons.append(dict(
        label=label,
        method="update",
        args=[{"visible": visible}, {"yaxis.title.text": label}],
    ))

fig.update_layout(
    barmode="group",
    title_text="Transfer throughput — higher is better",
    xaxis_title="Dataset size",
    yaxis_title="Upload (MB/s)",
    height=470,
    legend_title="Layout",
    template="plotly_white",
    updatemenus=[dict(
        type="buttons",
        x=0.0, y=1.1,
        xanchor="left",
        showactive=True,
        buttons=buttons,
    )],
)
fig.show()

## 4. Query Runtime

Use the **buttons** to switch between selective and broad query.
- **Selective**: filter `region=Zurich`, date `2026-01-10` to `2026-01-12` — partition pruning applies.
- **Broad**: full scan, no filter — all files must be opened.

by_region is **3.1x faster than flat** on selective. Flat is **7x faster than by_date** on broad.

In [14]:
query_metrics = [
    ("selective_time_s", "Selective query time (s)"),
    ("broad_time_s",     "Broad query time (s)"),
]

fig = go.Figure()

for m_idx, (col, label) in enumerate(query_metrics):
    for layout in LAYOUT_ORDER:
        subset = df[df["layout"] == layout].set_index("size").reindex(SIZES)
        fig.add_trace(go.Bar(
            name=layout,
            x=SIZES,
            y=subset[col],
            marker_color=COLORS[layout],
            text=subset[col].round(3).astype(str) + "s",
            textposition="outside",
            visible=(m_idx == 0),
            legendgroup=layout,
            showlegend=(m_idx == 0),
            hovertemplate=layout + " %{x}: %{y:.4f}s<extra></extra>",
        ))

n = len(LAYOUT_ORDER)
buttons = []
for m_idx, (col, label) in enumerate(query_metrics):
    visible = [False] * (len(query_metrics) * n)
    for i in range(n):
        visible[m_idx * n + i] = True
    buttons.append(dict(
        label=label,
        method="update",
        args=[{"visible": visible}, {"yaxis.title.text": label}],
    ))

fig.update_layout(
    barmode="group",
    title_text="Query runtime — lower is better",
    xaxis_title="Dataset size",
    yaxis_title="Selective query time (s)",
    height=470,
    legend_title="Layout",
    template="plotly_white",
    updatemenus=[dict(
        type="buttons",
        x=0.005, y=1.10,
        xanchor="left",
        showactive=True,
        buttons=buttons,
    )],
)
fig.show()

## 5. Scaling Behaviour

How does each layout scale as data grows from S to L?
Use the **buttons** to switch between selective and broad query scaling.
by_date shows super-linear growth on broad query due to file-count overhead.

In [17]:
row_counts = [SIZE_ROWS[s] for s in SIZES]

fig = go.Figure()

for m_idx, (col, label) in enumerate(query_metrics):
    for layout in LAYOUT_ORDER:
        subset = df[df["layout"] == layout].set_index("size").reindex(SIZES)
        fig.add_trace(go.Scatter(
            name=layout,
            x=row_counts,
            y=subset[col].values,
            mode="lines+markers",
            marker=dict(size=9),
            line=dict(color=COLORS[layout], width=2),
            visible=(m_idx == 0),
            legendgroup=layout,
            showlegend=(m_idx == 0),
            hovertemplate=layout + ": %{y:.3f}s<extra></extra>",
        ))

n = len(LAYOUT_ORDER)
buttons = []
for m_idx, (col, label) in enumerate(query_metrics):
    visible = [False] * (len(query_metrics) * n)
    for i in range(n):
        visible[m_idx * n + i] = True
    buttons.append(dict(
        label=label,
        method="update",
        args=[{"visible": visible}, {"yaxis.title.text": label}],
    ))

fig.update_layout(
    title_text="Query time scaling with dataset size",
    xaxis_title="Number of rows",
    xaxis_tickformat=",.0f",
    yaxis_title="Selective query time (s)",
    height=450,
    legend_title="Layout",
    template="plotly_white",
    updatemenus=[dict(
        type="buttons",
        x=0.005, y=1.1,
        xanchor="left",
        showactive=True,
        buttons=buttons,
    )],
)
fig.show()

## 6. Analysis Summary

In [18]:
print("\n=== ANALYSIS SUMMARY ===\n")

max_files = df.loc[df["local_file_count"].idxmax()]
print("1. Partitioned layouts create significantly more files than flat.")
print(f"   {max_files['size']}/by_date generates {int(max_files['local_file_count'])} files vs 1 for flat.")

worst_s3 = df.loc[df["s3_listing_time_s"].idxmax()]
worst_local = df.loc[
    (df["size"] == worst_s3["size"]) & (df["layout"] == worst_s3["layout"]),
    "local_listing_time_s"
].values[0]
ratio_listing = worst_s3["s3_listing_time_s"] / worst_local
print("\n2. S3 listing time scales badly with file count.")
print(f"   {worst_s3['size']}/{worst_s3['layout']}: {worst_s3['s3_listing_time_s']:.1f}s on S3 "
      f"vs {worst_local:.3f}s local (x{ratio_listing:.0f} overhead).")

l_sel = df[df["size"] == "L"].set_index("layout")["selective_time_s"]
ratio_sel = l_sel["flat"] / l_sel["by_region"]
print("\n3. Selective queries benefit most from by_region partitioning.")
print(f"   L: flat={l_sel['flat']:.2f}s, by_date={l_sel['by_date']:.2f}s, "
      f"by_region={l_sel['by_region']:.2f}s (x{ratio_sel:.1f} faster than flat).")

l_broad = df[df["size"] == "L"].set_index("layout")["broad_time_s"]
ratio_broad = l_broad["by_date"] / l_broad["flat"]
print("\n4. Broad queries favor flat layout (single sequential file read).")
print(f"   L: flat={l_broad['flat']:.2f}s, by_region={l_broad['by_region']:.2f}s, "
      f"by_date={l_broad['by_date']:.2f}s (x{ratio_broad:.1f} slower than flat).")

l_up = df[df["size"] == "L"].set_index("layout")["upload_throughput_mbs"]
ratio_up = l_up["flat"] / l_up["by_date"]
print("\n5. Transfer throughput is much higher for flat (fewer, larger files).")
print(f"   L upload: flat={l_up['flat']:.0f} MB/s vs by_date={l_up['by_date']:.1f} MB/s "
      f"(x{ratio_up:.0f} difference).")

print("\n=== RECOMMENDATIONS ===")
print("- Use by_region for selective/analytical queries filtering on a specific region.")
print("- Use flat for broad scans or when transfer speed matters.")
print("- Avoid by_date at large scale without compaction: S3 listing becomes prohibitive.")


=== ANALYSIS SUMMARY ===

1. Partitioned layouts create significantly more files than flat.
   L/by_date generates 3136 files vs 1 for flat.

2. S3 listing time scales badly with file count.
   L/by_date: 29.9s on S3 vs 0.191s local (x157 overhead).

3. Selective queries benefit most from by_region partitioning.
   L: flat=2.68s, by_date=1.03s, by_region=1.06s (x2.5 faster than flat).

4. Broad queries favor flat layout (single sequential file read).
   L: flat=1.76s, by_region=2.64s, by_date=8.20s (x4.7 slower than flat).

5. Transfer throughput is much higher for flat (fewer, larger files).
   L upload: flat=322 MB/s vs by_date=11.7 MB/s (x28 difference).

=== RECOMMENDATIONS ===
- Use by_region for selective/analytical queries filtering on a specific region.
- Use flat for broad scans or when transfer speed matters.
- Avoid by_date at large scale without compaction: S3 listing becomes prohibitive.
